### PR TITLE
chapter 12 vocabulary post

### PR DESCRIPTION
**Pull request by swemott**

### CODE REVIEW

**Comment by geanders** [comment]
@swemott : Great start on this! I have some suggestions below for small edits here and there, and a few places for you to go in and try to either make the definition more general or more specific. Once you make these on your computer, if you push them up to your GitHub, they should come right through in the pull request, as long as we keep it open (send me an email when you do just in case, so I can check). Once you do that, @baileyfosdick might have a few last edits to add directly in GitHub, and then we'll get your post up and let you know so you can make sure your Quizlet app uses the latest version of the vocab.

- I think that the definition for "predictors" might be a bit off. They're similar to independent variables in a regression context. For predictive models (which could follow a regression format), they have the added sense that they are variables / characteristics of each observation that might help in predicting the value of the outcome variable or target variable that you're interested in. So maybe change that definition to something like "characteristics measured for an observation that may be useful in predicting the target variable".
- For "objective function", could you move "in the context of supervised learning" from being part of the term to being the start of the definition?
- For "kernel methods" it would be helpful to also include a general idea of the class of what they are. Wikipedia has some nice language along these lines (https://en.wikipedia.org/wiki/Kernel_method). Maybe add to the start of the definition something like, "a class of algorithms for pattern analysis". Also, I recommend taking out the "do not need features at all" part. @baileyfosdick can correct me if I'm wrong about this, but I think these still do use the features (aka predictors or independent variables), but they then are using them to determine distances or similarity measures. The Wikipedia entry also has some other nice language to help define them that you might want to consider in your definition, something like "these use kernel functions, which enable them to operate in a high-dimensional, implicit feature space without ever computing the coordinates of the data in that space".
- For "classification", that was in last week's vocabulary, too. I don't think there's anything wrong with your definition here, but it might be nice to be consistent with them. I recommend that you replace the definition here with last week's definition: "the process of grouping observations in a dataset by their similarities in terms of measured characteristics".
- For "LDA", I think there's a typo. I think you're missing "in" between "used" and "supervised". Also, this definition is currently a bit generic and could describe several different methods. Perhaps adapt it a bit, to something like "a common technique used both for supervised learning classification and as a pre-processing dimension reduction step that finds a linear combination of features to help in classification." I think the "linear combination" helps in distinguishing this technique from some other possible approaches to classification problems, so it would be good to have that in the definition.
- For "misclassification rate", could you add something about that this is a measure for classification models? You wouldn't use this measure if you were trying to predict a continuous value (e.g., for a regression problem), just if you're trying to predict which class an observation belongs to.
- For "k-fold cross-validation", could you add at the beginning something like "a technique where". Also, at the end, could you explain what's done after making this split into testing and training data? You're using them to fit the model to the training portion and then assess, with measures like misclassification rate, how well it performs in the testing data (which was "held out" from being used to develop the model), so that ideally you can get a good gauge of how well the model does, including whether it seems to be overfitting to the exact data used to build the model. I think this end part of the technique is missing from the current definition.
- For "curse of dimensionality", could you add something to the end like "because data in any particular region becomes very sparse as dimensions increase"?
- For "confusion matrix", could you add to the end, "by counting the number of observations truly within each class versus the number predicted by the model to be in each class"?
- For "sensitivity" and "specificity", you've made the definitions a bit too specific to epidemiology. These terms are also used more generally to refer to misclassification of outcomes that can be marked as "positive" and "negative" (or have a binary outcome in general), regardless of whether the test is specifically testing for a positive/negative disease status. Could you revise these two definitions to make them more general? You could certainly use healthy/ill as examples, but I think we need to be clear that these are only one context in which you would use these terms.
- For "ROC curves", could you either change the term to end with "curve" or change the start of the definition to "graphical plots" so that the term and the definition agree in terms of singular / plural?
- For "bias", could you change the start of the definition from "measures" to "a measure of"?
- For "regularization", could you either add "methods" to the end of the term or change the start of the definition to "a method", so the term and definition will agree in singular / plural? Also, at the end of this definition, could you add "by helping to prevent overfitting of the model to the training data"?
- For "workflow", could you move "in the context of a computational analysis" from the end of the term to the start of the definition? Also, could you change "chains" to "the chaining of"?
- For "diagnostic plot" could you change the term to be plural ("diagnostic plots")? Also, I think this definition might be a bit too specific. I think you could use these plot to assess whether assumptions you made in fitting a certain model are reasonable, regardless of the type of model you're using (we certainly use them a lot in regression modeling, but I don't think that's the only type of model fitting where you'd use them). Could you see if you could make that definition a little more general?
- For "lasso", could you move "in the context of statistical regression modeling" from the end of the term to the start of the definition? Also, could you capitalize the starting letters in "Least Absolute Shrinkage and Selection Operator" to make it clearer that this is where we get "LASSO" from? Finally, I recommend that you change "helps ..." to "that is used in one type of regression modeling to reduce over-fitting and select useful features of the data for predicting the outcome."
- For "ridge regression", I suggest that you add to the beginning of the definition "a method of regression in which" and then changing the comma to a period and adding after that period "Doing this".
- For "penalty function", there's a typo ("contstraints").
- The definition of "logistic regression" is currently a bit too general---it could describe any algorithm for building a classification model (for example, it would also describe LDA or a classification tree model). Could you see if you could make that more specific and pick up what makes this algorithm different from other classification algorithms?
- For "tuning parameters", I suggest editing the definition a bit to: "parameters that control the strength of the penalty term in certain types of regression algorithms (e.g., ridge and lasso regression), controlling the amount of shrinkage (where parameter estimates are shrunk towards a central point, like the mean) when fitting the model."
- For "objective function", I recommend that you change the definition to something like "the function that you optimize during the training of a predictive model (e.g., the maximum likelihood function for a classic regression model)". For an interesting discussion, see https://stats.stackexchange.com/questions/179026/objective-function-cost-function-loss-function-are-they-the-same-thing

**Comment by geanders** [comment]
@swemott : Great! I've now got that post up on our website: https://kind-neumann-789611.netlify.app/post/vocabulary-for-chapter-12/

I'm going to leave this pull request open for you to submit your exercise post through. I think you're still working on that, right?

**Comment by swemott** [comment]
Thank you, Brooke. 

I have to admit, I’m struggling to find the time to work through the exercise post. I’m typically a lot more on top of things, but this whole having my kiddo home 24/7 has definitely thrown me for a loop. I will do my best to work on it this weekend and let you know if I have any questions. My apologies for the delay. 

-Sherry

**Comment by geanders** [comment]
No problem at all! Please let me know if you'd like to talk offline to figure out something that will be reasonable for your situation.

**Comment by geanders** [comment]
@swemott : It looks like this came through fine! I'll take a look tomorrow and add some suggestions for next steps and edits.

**Comment by geanders** [comment]
@swemott : This is a good start, and I think you're on the right track. Here are some suggestions for moving forward: 

- Right now, it's outputting to a pdf instead of a blog post. We should be able to copy everything into a blog post format once you're done, though (or you could try to create a new blog post---same way you did for the vocab---and then copy all of the main text from your exercise into that).
- When you show the `PlantGrowth` data, you probably don't need to show all of it. A function like `head` or `slice` might be helpful in showing just a bit.  
- After the sentence, "We are instructed to make a less extreme example of correlated test statistics than the data duplication at the end of Section 6.5.", could you say that you'll be explaining and showing what we did for that data duplication later in this blog post? You end up having some code that walks through that, so it would be helpful to let the reader know you'll be covering that, too (and they don't necessarily need to go back and revisit it themselves to understand this post). 
- For the very first paragraph, some people have been copying the exercise prompt directly from the text book and then putting it in the blog post in a "quotation" environment (just add "> " at the start of the paragraph to do that in RMarkdown). If you wanted, you could edit the first paragraph to do that. 
- There's a typo: "yeilds"
- For any R object name or package name, could you add backticks around it so it will show up in code? Some examples of places without that formatting where it would be nice include: ctrl, trt1, and trt2. This extra formatting will help readers, because they will quickly be able to distinguish that those words aren't regular English words and their brains don't need to try to parse them as English.
- There's another typo: "taht"
- For the text, "Here we resample only half the data:", could you add ", so the size of the data set is the same as the original data, but now half of the observations are fully dependent (exactly the same) as other observations in the dataset". 
- For the paragraph where you have the text "Next we'll apply a t-test comparing weights in the ctrl group to the trt2 group:", could you add a sentence or so to explain the null hypothesis that this t-test is testing against? (that the means of the weights of the control and treatment 2 groups are equal, I'm guessing). 
- Overall, I think it would be useful to include a bit of discussion about the point of this exercise. When you do a t-test (for example, using `t.test`), you make the assumption that the observations are independent from each other. The p-value that you get from a t-test could be way off if you violate this assumption in a big way. This might be a point to do a touch more research on (e.g., googling "problems with violating independent assumption t-test" or something like that) to help in writing up that discussion and explaining the overarching question we're trying to check out in this exercise (and in the exercise in the main part of the text that it builds off of). 
- It might be nice to do something at the end of the post to summarize and pull together what you've gotten from these different datasets you've created. You could maybe make a table where you have one column that gives a description of each of the datasets you created (original, doubled without any random noise, same size as original but with half the observations replication, same but with some random noise added to each replicate), then a column with "N" (the number of datapoints in that dataset), then columns with the t-test statistics and the p-values that you get from applying the t-test to each of these datasets. Then you could talk a bit about how the test statistic and p-value seem to be influenced by (a) the size of the data and (b) violations of the independence assumptions in the data.

**Comment by geanders** [comment]
@swemott : Really nice work revising this! I have some minor editing and code suggestions that I've listed below. I think the only big picture things to do next are (1) to set a seed early in the code, so your results with sampled data won't change on you each time this document is re-rendered and (2) do some more to explain why you're adding in the permutation test and how to interpret the results from that.

- Really nice job of adding text explaining steps along the way!
- I suggest changing "by a two-sample paired test" to "(e.g., by a two-sample paired test for paired data", just to make sure it's clear that this isn't an approach you'd want to use if the data were duplicated (but the point about paired data is totally valid here---I think that's a good example, just wanted to clarify in the text). 
- For the text "Here we resample only half the data;", I recommend you add "and then duplicated that sampled half to create a dataset".
- Because you're using random samples, the p-value is going to change from time to time for any result from duplicated data. Therefore, if you write a p-value from one run into the text, it might not be the same as in the next run. Because of this, you might want to change the text "The p-value increases from 0.004 with a fully duplicated dataset to 0.423 with half the data duplicated." to end instead with "to a value typically much higher with half the data duplicated" or something else that's general enough that it will still make sense as you get different random samples with knitting the document different times. (Another approach would be to set a seed, so you force it to give you the same random sample each time.)
- Add a blank line in the Rmd before the text that starts: "Now we’ll take the dataset with half ". That will make sure that the new paragraph doesn't start in the same space (same line) as the figure from the previous code chunk was printed. (In general, it's a good idea to always leave a blank line on each side of each code chunk.)
- After "Now we’ll take the dataset with half resampled, and add random noise to each observation.", I suggest that you add something like: "The following code uses the `rnorm` function to add random noise to each value from a normal distribution centered at 0 and with a standard deviation of 0.2".
- For the permutation tests, I'm guessing that you're running those because they don't require this assumption of independence, and so will work even with these duplicated datasets? You might want to revisit in the book why we might want to apply permutation tests, and add just a line or two of text explaining that (I would suggest replacing "Permutation test data resample to check Type 1 error control:" with more explanation of why you're doing this step), and then also check on how we should be interpreting the objects like `pg1_null` you get from doing that, as well as the histogram you're printing, and add a bit of text walking the reader how to look through this output and what they can figure out from it. (also, it might help to print out the output from the permutation test (`pg1_null`, for example). 
- When you run the first permutation test, you're using a dataset that's a bit different that the one you just showed. With the code: 

```
pg1 <- PlantGrowth %>% 
  sample_frac(size = 0.5) 
```

you're randomly sampling half of the data, but not doing that next step of then duplicating this half. If you want to use the same size data as before, you could change this to: 

```
pg1 <- PlantGrowth %>% 
  sample_frac(size = 0.5) %>%
  bind_rows(., .)
```

That `bind_rows` is sticking together two copies of the sampled data. Alternatively, you could go ahead and create that `pg1` object earlier. Then your code in the previous chunk would look more like: 

```
pg1 <- PlantGrowth %>% 
  sample_frac(size = 0.5) %>% 
  bind_rows(., .) 

pg1 %>% 
  filter(group %in% c("ctrl", "trt2")) %>% 
  t.test(weight ~ group, data  = .) %>% 
  tidy()
```

and then in this second chunk could go straight into: 

```
pg1_null = with(
  dplyr::filter(pg1, group %in% c("ctrl", "trt2")),
    replicate(10000,
      abs(t.test(weight ~ sample(group))$statistic)))

ggplot(tibble(`|t|` = pg1_null), aes(x = `|t|`)) +
  geom_histogram(binwidth = 0.1, boundary = 0)
```

If you do this, then you can still create `pg2` in the same way, but you won't need to bind the rows of `pg1` and `pg2` before you do the t-test, because you'll already be working with the right size dataset. It's fine to add random noise to *all* the observations, both the original sample and the duplicated. Your code for `pg2` would then be: 

```
pg2 <- pg1 %>% 
  mutate(noise = rnorm(15, mean = 0, sd = 0.2), 
         weight = weight + noise) %>% 
  select(-noise)
pg2

pg2 %>% 
  filter(group %in% c("ctrl", "trt2")) %>% 
  t.test(weight ~ group, data  = .) %>% 
  tidy()
```

- In the chunk where you load `kableExtra`, you might want to set the chunk options `message = FALSE, warning = FALSE` so it's won't print out the stuff about masking in the post output. 
- Nice table! The formatting's really great. However, since you're randomly sampling, it won't work to put in the test statistics and p-values in "by hand" (as you do with `test_statistic <- c(-2.13, -3.1, -3.03, -0.72)` and `p_value <- c(0.479, 0.003, 0.009, 0.48)`) because they'll change every time you render the post. One option is to set the seed at the start of the post (e.g., with something like `set.seed(1234)`, and then get the numbers from that run. If you set the seed, they'll always be the same. The other would be to assign the output from the t.test for each dataset to an R object and then extract the p-values from those objects in this code. In the short term, setting the seed would probably be the quicker option. 
- Somewhere, you probably want to make a note that, because these are random samples, the exact p-values are variable, and if someone runs this code with a different seed, they'll get different results. 
- In the table, I think that you've got a typo for the p-value for the original data's t-test. I think that should be closer to "0.049", rather than "0.49".

**Comment by geanders** [comment]
Now published: https://kind-neumann-789611.netlify.app/post/exercise-solution-for-chapter-6/
